Libraries/src/net/relinc/libraries/application/LineChartWithMarkers.java

```diff
@@ -6,182 +6,421 @@
 
-            Rectangle rectangle = new Rectangle(0,0,0,0);
-            rectangle.setStroke(Color.TRANSPARENT);
-            rectangle.setFill(color.deriveColor(1, 1, 1, 0.2));
-            
-            marker.setNode(rectangle);
+        for (Data<X, X> verticalRangeMarker : verticalRangeMarkers) {
+
+            Rectangle rectangle = (Rectangle) verticalRangeMarker.getNode();
+            rectangle.setX(getXAxis().getDisplayPosition(verticalRangeMarker.getXValue()) + 0.5);  // 0.5 for crispness
+            rectangle.setWidth(getXAxis().getDisplayPosition(verticalRangeMarker.getYValue()) - getXAxis().getDisplayPosition(verticalRangeMarker.getXValue()));
+            rectangle.setY(0d);
+            rectangle.setHeight(getBoundsInLocal().getHeight());
+            rectangle.toBack();
 
-            getPlotChildren().add(rectangle);
-            verticalRangeMarkers.add(marker);
-            setMouseTransparentToEverythingButBackground();
         }
-        
-        public void setMouseTransparentToEverythingButBackground(){
-        	final Node chartBackground = lookup(".chart-plot-background");
-            for (Node n: chartBackground.getParent().getChildrenUnmodifiable()) {
-              if (n != chartBackground) {
+
+    }
+
+    public void addVerticalRangeMarker(Data<X, X> marker, Color color) {
+        Objects.requireNonNull(marker, "the marker must not be null");
+        if (verticalRangeMarkers.contains(marker)) return;
+
+        Rectangle rectangle = new Rectangle(0, 0, 0, 0);
+        rectangle.setStroke(Color.TRANSPARENT);
+        rectangle.setFill(color.deriveColor(1, 1, 1, 0.2));
+
+        marker.setNode(rectangle);
+
+        getPlotChildren().add(rectangle);
+        verticalRangeMarkers.add(marker);
+        setMouseTransparentToEverythingButBackground();
+    }
+
+    public void setMouseTransparentToEverythingButBackground() {
+        final Node chartBackground = lookup(".chart-plot-background");
+        for (Node n : chartBackground.getParent().getChildrenUnmodifiable()) {
+            if (n != chartBackground) {
                 n.setMouseTransparent(true);
-              }
             }
         }
+    }
 
-        public void removeVerticalRangeMarker(Data<X, X> marker) {
-            Objects.requireNonNull(marker, "the marker must not be null");
-            if (marker.getNode() != null) {
-                getPlotChildren().remove(marker.getNode());
-                marker.setNode(null);
-            }
-            verticalRangeMarkers.remove(marker);
+    public void removeVerticalRangeMarker(Data<X, X> marker) {
+        Objects.requireNonNull(marker, "the marker must not be null");
+        if (marker.getNode() != null) {
+            getPlotChildren().remove(marker.getNode());
+            marker.setNode(null);
         }
-        
+        verticalRangeMarkers.remove(marker);
+    }
+
+    public void chartWithPreferences(ChartingPreferences preference, String defaultTitle) {
```

Comment:
Can we rename this method to `applyPreferences`? I think the `apply` verb is helpful to show that it is mutating this chart

Reply:
that seems more appropriate